server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -283,12 +283,17 @@ public RegisterRepositoryTask(final RepositoriesService repositoriesService, fin
 
         @Override
         public ClusterState execute(ClusterState currentState) {
-            RepositoryMetadata newRepositoryMetadata = new RepositoryMetadata(request.name(), request.type(), request.settings());
             Metadata.Builder mdBuilder = Metadata.builder(currentState.metadata());
             RepositoriesMetadata repositories = RepositoriesMetadata.get(currentState);
             List<RepositoryMetadata> repositoriesMetadata = new ArrayList<>(repositories.repositories().size() + 1);
             for (RepositoryMetadata repositoryMetadata : repositories.repositories()) {
-                if (repositoryMetadata.name().equals(newRepositoryMetadata.name())) {
+                if (repositoryMetadata.name().equals(request.name())) {
+                    final RepositoryMetadata newRepositoryMetadata = new RepositoryMetadata(
+                        request.name(),
+                        repositoryMetadata.uuid(),
```

Comment:
Here we copy the UUID from the previous repository instance rather than using `_na_`. The next time we load the `RepositoryData` we update the metadata if needed:

https://github.com/elastic/elasticsearch/blob/cef1b54506e05040cc7a000cce54bdd1df90c254/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java#L2465-L2475

We could conceivably be stricter here, see https://github.com/elastic/elasticsearch/issues/109936, but it doesn't seem necessary today. Instead note that in `RepositorySupplier` we'll notice the change in UUID and look for a different repository with matching UUID before eventually throwing a `RepositoryMissingException`.

Reply:
I understand this is an optimisation when UUID does not change which should be most of the cases.

When the UUID does change, are you saying that 
1. Copying the UUID only delays the need to load repositoryData.
2. The repositoryData will be loaded before any writing, e.g. createSnapshot, can happen so that UUID will become consistent.

It is not clear to me why a cached repositoryData would not be loaded in 2 and further delays the UUID consistency update?

The change somehow feels not belong here. But I may be too paranoid.

Reply:
> create a brand-new Repository

Thanks! This is an important information that I originally missed. 

> If we don't do this then we can't see that the new Repository instance is the one we should use for searchable snapshot operations in future (at least not without blocking some thread somewhere while waiting for the new UUID to be loaded).

I see the point now. I guess that means searchable snapshot actions do not always load repo data as the first step? If the repo UUID did change, does that mean it would take a while before searchable snapshot related code realise it? Would that lead to any issue?

Reply:
> I guess that means searchable snapshot actions do not always load repo data as the first step?

Searchable snapshot actions essentially never load the `RepositoryData`. They already know how to find the shard data within the blob store (from the `index.store.snapshot.index_uuid` and `index.store.snapshot.snapshot_uuid` settings in the index metadata, and the shard ID). If the repo switches out from underneath them then they'll get exceptions indicating that the blobs they need are no longer found.